### ISSUE TITLE
setting RPC header and api_base

### ISSUE THREAD

**Issue by josephykwang** [gepa-ai/gepa#45]
we need to set RPC headers and LLM gateway. The current adapter examples don't seem to support these.

**Comment by LakshyAAAgrawal** [comment]
Dear @josephykwang ,

Could you please share an example for the interface you would like to see supported? Is this for litellm?

**Comment by josephykwang** [comment]
I'm using the example on the GitHub ```# Load AIME dataset
trainset, valset, _ = gepa.examples.aime.init_dataset()

seed_prompt = {
    "system_prompt": "You are a helpful assistant. You are given a question and you need to answer it. The answer should be given at the end of your response in exactly the format '### <final answer>'"
}

# Let's run GEPA optimization process.
gepa_result = gepa.optimize(
    seed_candidate=seed_prompt,
    trainset=trainset, valset=valset,
    task_lm="openai/gpt-4.1-mini", # <-- This is the model being optimized
    max_metric_calls=150, # <-- Set a budget
    reflection_lm="openai/gpt-5", # <-- Use a strong model to reflect on mistakes and propose better prompts
)

print("GEPA Optimized Prompt:", gepa_result.best_candidate['system_prompt'])``` think you are using litellm within gepa/api.py.


we need to pass it token, set RPC headers, and set api_base to our LLM gateway.

**Comment by LakshyAAAgrawal** [comment]
Hey! Please take a look at [https://github.com/gepa-ai/gepa/blob/d39e4d0979f4f7e1d382343c4e761b8067e6be6a/src/gepa/adapters/default_adapter/default_adapter.py#L25](https://github.com/gepa-ai/gepa/blob/d39e4d0979f4f7e1d382343c4e761b8067e6be6a/src/gepa/adapters/default_adapter/default_adapter.py#L25). You can provide a custom function, that will be called with the `messages` arg ([https://github.com/gepa-ai/gepa/blob/main/src/gepa/adapters/default_adapter/default_adapter.py#L65](https://github.com/gepa-ai/gepa/blob/main/src/gepa/adapters/default_adapter/default_adapter.py#L65)). You can set whatever parameters you want when calling it with that.

You can set the `task_lm` parameter in the optimize method (https://github.com/gepa-ai/gepa/blob/main/src/gepa/api.py#L25). Please let me know if this doesn't resolve your query, and feel free to reopen the issue!

**Comment by josephykwang** [comment]
try ```import gepa
import openai
from openai import OpenAI

def llm_call(messages):
    openai.api_base = '....'
    openai.api_key = '....'

    headers = {
      ...
    }

    client = OpenAI(base_url=openai.api_base, api_key=openai.api_key, default_headers=headers)
    
    response = client.chat.completions.create(
        model="gpt-4.1-mini",
        messages=messages
    )
    return response.choices[0].message.content.strip()

class ReflectionLM:
    def __call__(self, prompt: str) -> str:
        openai.api_base = '...'
        openai.api_key = '...'

        headers = {
            ...
        }

        client = OpenAI(base_url=openai.api_base, api_key=openai.api_key, default_headers=headers)
        
        response = client.chat.completions.create(
            model="gpt-5",
            messages=[{"role": "user", "content": prompt}]
        )
        return response.choices[0].message.content.strip()

# Load AIME dataset
trainset, valset, _ = gepa.examples.aime.init_dataset()

seed_prompt = {
    "system_prompt": "You are a helpful assistant. You are given a question and you need to answer it. The answer should be given at the end of your response in exactly the format '### <final answer>'"
}

reflection_lm = ReflectionLM()

# Let's run GEPA optimization process with custom LLM implementations
gepa_result = gepa.optimize(
    seed_candidate=seed_prompt,
    trainset=trainset, 
    valset=valset,
    task_lm=llm_call,          
    max_metric_calls=50,           
    reflection_lm=reflection_lm,    
)

print("GEPA Optimized Prompt:", gepa_result.best_candidate['system_prompt'])
```  it outputs: ```GEPA Optimized Prompt: You are a helpful assistant. You are given a question and you need to answer it. The answer should be given at the end of your response in exactly the format '### <final answer>'
``` doesn't seem to optimize.

**Comment by josephykwang** [comment]
1) what is the min # of records needed in trainset and valet to apply gepa?
2) what is the min metric_calls?

**Comment by josephykwang** [comment]
Iteration 0: Base program full valset score: 0.26666666666666666
...
Iteration 3: Selected program 1 score: 0.3111111111111111

**Comment by josephykwang** [comment]
it seems we cannot get consistent improvement when rerunning the same code. any idea?

**Comment by LakshyAAAgrawal** [comment]
> what is the min # of records needed in trainset and valet to apply gepa?

We have seen results improve with as small as 3 examples in trainset (for example, https://dspy.ai/tutorials/gepa_papillon/). The typical recommendation is: Keep valset as small as possible, while ensuring it matches the final task/test distribution as close as possible (valset is only used to select the best programs). Use all the remaining data that you have as the training set.

> what is the min metric_calls?

As shared above, we have seen it improve with as few as 10 metric_calls. For most tasks, we recommend you set this to at least 20-30x of valset size.

**Comment by LakshyAAAgrawal** [comment]
> it seems we cannot get consistent improvement when rerunning the same code. any idea?

GEPA is a stochastic optimizer, so across runs, there will be some stochasticity. However, over a long enough optimization run, this should average out. Can you share more details about your runs?

**Comment by josephykwang** [comment]
one use case it to figure out what additional context to include in the prompt. any suggestion how to leverage gepa to select min set of context that will give the best overall accuracy?

also, pointers for using customized eval metrics?
